File: src/commands/branch.js
import { simpleGit } from "simple-git";

const git = simpleGit();

export default function branchCmdsprogram) {
    program
        .command("branch <name>")
        .description("Switch to an existing branch or create a new one")
        .action(async (name) => {
            try {
                const branchSummary = await git.branchLocal();

                if (branchSummary.all.includes(name)) {
                    global.log.info(`Switching to existing branch: ${name}...`);
                    await git.checkout(name);
                } else {
                    global.log.info(`Creating and switching to new branch: ${name}...`);
                    await git.checkoutLocalBranch(name);
                }

                global.log.info(`Successfully switched to branch "${name}".`);
            } catch (error) {
                global.log.error("Error switching branch:");
                global.log.error(error.message);
            }
        });
}

File: cmd/branch.go
package cmd

import (
	"github.com/ramadanny/gits/internal/gitops"
	"github.com/ramadanny/gits/internal/logger"
	"github.com/spf13/cobra"
)

func init() {
	rootCmd.AddCommand(&cobra.Command{
		Use:   "branch [name]",
		Short: "List branches, switch to an existing branch, or create a new one",
		Args:  cobra.MaximumNArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			if len(args) == 0 {
				out, _ := gitops.Run("branch")
				logger.Info("\n" + out)
				return
			}

			name := args[0]
			_, err := gitops.Run("rev-parse", "--verify", name)

			var checkoutErr error
			if err == nil {
				logger.Info("Switching to existing branch: " + name + "...")
				_, checkoutErr = gitops.Run("checkout", name)
			} else {
				logger.Info("Creating and switching to new branch: " + name + "...")
				_, checkoutErr = gitops.Run("checkout", "-b", name)
			}

			if checkoutErr != nil {
				logger.Error("Failed to switch branch. Make sure you don't have uncommitted changes that conflict.")
			} else {
				logger.Info("Successfully switched to branch \"" + name + "\".")
			}
		},
	})
}